tests/lib/languages/js/source-code/source-code.js

```diff
@@ -3606,4 +3571,680 @@ describe("SourceCode", () => {
+										);
+										assert.isFalse(
+											sourceCode.isGlobalReference(
+												stringRef,
+											),
+											"Imported 'String' should not be identified as a global reference",
+										);
+									});
+
+									return { "Program:exit": spy };
+								},
+							},
+						},
+					},
+				},
+				rules: { "test/checker": "error" },
+			};
+
+			flatLinter.verify(code, config);
+			assert(spy && spy.calledOnce, "Spy was not called.");
+		});
+
+		it("should handle temporal dead zone (TDZ) for let variables", () => {
+			const code = "{ console.log(x); let x = 5; }";
+			let spy;
+
+			const config = {
+				plugins: {
+					test: {
+						rules: {
+							checker: {
+								create(context) {
+									const sourceCode = context.sourceCode;
+
+									spy = sinon.spy(node => {
+										if (
+											node.name === "x" &&
+											node.parent.type !==
+												"VariableDeclarator"
+										) {
+											assert.isFalse(
+												sourceCode.isGlobalReference(
+													node,
+												),
+												"Reference in TDZ should not be identified as a global reference",
+											);
+										}
+									});
+
+									return { Identifier: spy };
+								},
+							},
+						},
+					},
+				},
+				rules: { "test/checker": "error" },
+				languageOptions: { ecmaVersion: 2015 },
+			};
+
+			flatLinter.verify(code, config);
+			assert(spy && spy.called, "Spy was not called.");
+		});
+
+		it("should handle variables shadowed in catch blocks", () => {
+			const code = "try {} catch (Error) { Error; }";
+			let spy;
+
+			const config = {
+				plugins: {
+					test: {
+						rules: {
+							checker: {
+								create(context) {
+									const sourceCode = context.sourceCode;
+
+									spy = sinon.spy(node => {
+										if (
+											node.parent.type === "CatchClause"
+										) {
+											// Skip the catch parameter declaration
+											return;
+										}
+
+										if (node.name === "Error") {
+											assert.isFalse(
+												sourceCode.isGlobalReference(
+													node,
+												),
+												"Error in catch block should not be identified as a global reference",
+											);
+										}
+									});
+
+									return { Identifier: spy };
+								},
+							},
+						},
+					},
+				},
+				rules: { "test/checker": "error" },
+			};
+
+			flatLinter.verify(code, config);
+			assert(spy && spy.called, "Spy was not called.");
+		});
+
+		it("should handle class declarations and methods", () => {
+			const code =
+				"class MyClass { method() { Math.random(); this.Math = 5; } }";
+			let spy;
+
+			const config = {
+				plugins: {
+					test: {
+						rules: {
+							checker: {
+								create(context) {
+									const sourceCode = context.sourceCode;
+
+									spy = sinon.spy(node => {
+										if (
+											node.name === "Math" &&
+											node.parent.type ===
+												"MemberExpression" &&
+											!node.parent.object.type ===
```

Comment:
```suggestion
											node.parent.object.type !==
```

Othwerise, this would never be `true`.